### PR TITLE
Fix Python service issues in gog.py

### LINKED ISSUE #6686: Install DLC or updates of some games by GOG
### Bug description

When you click on install DLCs or updates of some games, in my case it is Europa Universalis IV, a 404 error message is displayed.

### How to Reproduce

I only click on installing DLCs or updates.

### Expected behavior

Get updates and game DLCs.

### Log output

```shell
INFO     2026-05-11 01:32:01,364 [application.do_command_line:468]:Starting Lutris 0.5.22
WARNING: [Loader Message] Code 0 : terminator_CreateInstance: Received return code -9 from call to vkCreateInstance in ICD /usr/lib64/libvulkan_dzn.so. Skipping this driver.
WARNING: [Loader Message] Code 0 : ICD for selected physical device does not export vkGetPhysicalDeviceDisplayPlanePropertiesKHR!
WARNING: [Loader Message] Code 0 : ICD for selected physical device does not export vkGetPhysicalDeviceDisplayPropertiesKHR!
WARNING: [Loader Message] Code 0 : terminator_CreateInstance: Received return code -9 from call to vkCreateInstance in ICD /usr/lib64/libvulkan_dzn.so. Skipping this driver.
WARNING: [Loader Message] Code 0 : ICD for selected physical device does not export vkGetPhysicalDeviceDisplayPlanePropertiesKHR!
WARNING: [Loader Message] Code 0 : ICD for selected physical device does not export vkGetPhysicalDeviceDisplayPropertiesKHR!
INFO     2026-05-11 01:32:01,756 [startup.run_all_checks:120]:"card1" is AMD Radeon RX 550 / 550 Series (1002:699f 1002:0b04 amdgpu) Driver 25.3.6
DEBUG    2026-05-11 01:32:02,095 [path_cache._update_missing_games:114]:Checking for missing games
DEBUG    2026-05-11 01:32:02,217 [api.check_stale_runtime_versions:52]:Modified at seg 11 mai 2026 01:09:59, will update after seg 11 mai 2026 07:09:59
DEBUG    2026-05-11 01:32:02,219 [lutriswindow.create_runtime_updater_cb:1460]:Runtime up to date
DEBUG    2026-05-11 01:32:03,099 [http._request:119]:POST https://lutris.net/api/users/library?since=1778472601
DEBUG    2026-05-11 01:32:06,423 [http._request:119]:GET https://embed.gog.com/user/data/games
INFO     2026-05-11 01:32:06,839 [gog.get_game_details:289]:Getting game details for 2057001589
DEBUG    2026-05-11 01:32:06,855 [http._request:119]:GET https://api.gog.com/products/2057001589?expand=downloads&locale=en-US
DEBUG    2026-05-11 01:32:07,376 [http._request:119]:GET https://api.gog.com/products?ids=1078355380,1085117592,1096200900,1121252700,1123779969,1127529116,1148881716,1157668974,1160224563,1167512084,1184841315,1205371409,1212570569,1215235007,1241813716,1258950752,1261075972,1263421402,1300580999,1318893118,1327205250,1329600501,1370930711,1389473830,1401170030,1414238826,1416197025,1434010780,1436373349,1492888940,1501330572,1513629651,1527338467,1550503498,1569220056,1570624433,1593401624,1593977772,1596121849,1598364621,1601780111,1607813411,1645948296,1684501525,1684867724,1708077552,1733507797,1740542108,1762534164,1766475832,1775857033,1780428699,1790798501,1795881620,1804554641,1812542500,1813286041,1819292625,1829098047,1877971339,1892718327,1918750011,1935124337,1940550307,1964362003,1979245360,2006763027,2016191934,2021140455,2045765337,2062585115,2142991031,2145664590&expand=downloads
ERROR    2026-05-11 01:32:07,850 [jobs.target:64]:Error while completing task <bound method GOGService.get_dlc_installers_runner of <lutris.services.gog.GOGService object at 0x7fc35ff2c050>>: <class 'lutris.util.http.HTTPError'> HTTP Error 400: Bad Request
  File "/usr/lib/python3.14/site-packages/lutris/util/jobs.py", line 62, in target
    result = self.function(*a, **kw)
  File "/usr/lib/python3.14/site-packages/lutris/services/gog.py", line 713, in get_dlc_installers_runner
    installers = self.get_dlc_installers_owned(db_game)
  File "/usr/lib/python3.14/site-packages/lutris/services/gog.py", line 703, in get_dlc_installers_owned
    installers = self.get_dlc_installers(db_game)
  File "/usr/lib/python3.14/site-packages/lutris/services/gog.py", line 649, in get_dlc_installers
    dlcs = self.get_game_dlcs(appid)
  File "/usr/lib/python3.14/site-packages/lutris/services/gog.py", line 283, in get_game_dlcs
    return self.make_api_request(all_products_url)
           ~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.14/site-packages/lutris/services/gog.py", line 235, in make_api_request
    request.get()
    ~~~~~~~~~~~^^
  File "/usr/lib/python3.14/site-packages/lutris/util/http.py", line 168, in get
    return self._request("GET", data)
           ~~~~~~~~~~~~~^^^^^^^^^^^^^
  File "/usr/lib/python3.14/site-packages/lutris/util/http.py", line 132, in _request
    raise HTTPError("%s" % error, code=error.code) from error
ERROR    2026-05-11 01:32:07,856 [exception_backstops.error_wrapper:86]:Error handling timeout function: HTTP Error 400: Bad Request
Traceback (most recent call last):
  File "/usr/lib/python3.14/site-packages/lutris/util/http.py", line 126, in _request
    request = self.opener.open(req, timeout=self.timeout)
  File "/usr/lib64/python3.14/urllib/request.py", line 493, in open
    response = meth(req, response)
  File "/usr/lib64/python3.14/urllib/request.py", line 602, in http_response
    response = self.parent.error(
        'http', request, response, code, msg, hdrs)
  File "/usr/lib64/python3.14/urllib/request.py", line 531, in error
    return self._call_chain(*args)
           ~~~~~~~~~~~~~~~~^^^^^^^
  File "/usr/lib64/python3.14/urllib/request.py", line 464, in _call_chain
    result = func(*args)
  File "/usr/lib64/python3.14/urllib/request.py", line 611, in http_error_default
    raise HTTPError(req.full_url, code, msg, hdrs, fp)
urllib.error.HTTPError: HTTP Error 400: Bad Request

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/lib/python3.14/site-packages/lutris/exception_backstops.py", line 84, in error_wrapper
    return handler(*args, **kwargs)
  File "/usr/lib/python3.14/site-packages/lutris/util/jobs.py", line 126, in wrapper
    func(*a, **kw)
    ~~~~^^^^^^^^^^
  File "/usr/lib/python3.14/site-packages/lutris/util/busy.py", line 36, in on_completion
    callback(*a, **kw)
    ~~~~~~~~^^^^^^^^^^
  File "/usr/lib/python3.14/site-packages/lutris/game.py", line 414, in on_installers_ready
    raise error  # bounce errors off the backstop
    ^^^^^^^^^^^
  File "/usr/lib/python3.14/site-packages/lutris/util/jobs.py", line 62, in target
    result = self.function(*a, **kw)
  File "/usr/lib/python3.14/site-packages/lutris/services/gog.py", line 713, in get_dlc_installers_runner
    installers = self.get_dlc_installers_owned(db_game)
  File "/usr/lib/python3.14/site-packages/lutris/services/gog.py", line 703, in get_dlc_installers_owned
    installers = self.get_dlc_installers(db_game)
  File "/usr/lib/python3.14/site-packages/lutris/services/gog.py", line 649, in get_dlc_installers
    dlcs = self.get_game_dlcs(appid)
  File "/usr/lib/python3.14/site-packages/lutris/services/gog.py", line 283, in get_game_dlcs
    return self.make_api_request(all_products_url)
           ~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.14/site-packages/lutris/services/gog.py", line 235, in make_api_request
    request.get()
    ~~~~~~~~~~~^^
  File "/usr/lib/python3.14/site-packages/lutris/util/http.py", line 168, in get
    return self._request("GET", data)
           ~~~~~~~~~~~~~^^^^^^^^^^^^^
  File "/usr/lib/python3.14/site-packages/lutris/util/http.py", line 132, in _request
    raise HTTPError("%s" % error, code=error.code) from error
lutris.util.http.HTTPError: HTTP Error 400: Bad Request
```

### System Information

```shell
Vulkan support: YES
Esync support: YES
Fsync support: YES
Wine installed: YES
Gamescope: YES
Mangohud: NO
Gamemode: YES
Steam: YES
In Flatpak: NO
[System]
OS: Fedora Linux 43 Workstation Edition
Arch: x86_64
Kernel: 6.19.14-200.fc43.x86_64
Lutris Version: 0.5.22
Python Version: 3.14.4 (main, Apr 16 2026, 00:00:00) [GCC 15.2.1 20260123 (Red Hat 15.2.1-7)]
Desktop: GNOME
Display Server: wayland
[CPU]
Vendor: GenuineIntel
Model: Intel(R) Core(TM) i7-3770 CPU @ 3.40GHz
Physical cores: 4
Logical cores: 8
[Memory]
RAM: 15.6 GB
Swap: 8.0 GB
[Graphics]
Vendor: AMD
OpenGL Renderer: AMD Radeon RX 550 / 550 Series (radeonsi, polaris12, ACO, DRM 3.64, 6.19.14-200.fc43.x86_64)
OpenGL Version: 4.6 (Compatibility Profile) Mesa 25.3.6
OpenGL Core: 4.6 (Core Profile) Mesa 25.3.6
OpenGL ES: OpenGL ES 3.2 Mesa 25.3.6
Vulkan Version: 1.4.341
Vulkan Drivers: AMD Radeon RX 550 / 550 Series (RADV POLARIS12) (1.4.328)
```

### Media (optional)

_No response_

### PR DESCRIPTION
**Pull request by KTGX710**
Fixes #6686:
Found issue that API request with ids parameter greater than 50 throws 403 error.
Developed iterations to break down excessive requests into multiple parts and recompile JSON afterward.
Only executes if ids' are more than 50, otherwise defaults to original logic.

Was unable to implement and run my unit/integration tests due to incessant dependency errors. Spent hours installing all packages listed in INSTALL.rts file but ultimately still unable to get working. dbus and lxml failed to be imported by my pip installation.

Any feedback on this issue, or my code and test development would be much appreciated.

### CODE REVIEW

**Comment by ItsAllAboutTheCode** [comment]
If you need an example of how to setup your development you can look at the github workflow [static.yml](https://github.com/lutris/lutris/blob/master/.github/workflows/static.yml#L103-L117) file.

I recommended creating a python virtual environment first, then using the `make req-python` and `make dev` [Makefile](https://github.com/lutris/lutris/blob/master/Makefile#L92-L101) targets to install the python dependencies into that virtual environment.

Either way, after you setup the development dependencies, you will need to [make format](https://github.com/lutris/lutris/blob/master/Makefile#L110) to fix the ruff formatting issues with your code.

File: lutris/services/gog.py [hunk 1]
```diff
@@ -282,10 +282,51 @@ def get_products_page(self, page: int = 1, search: str | None = None) -> Any:
...
+        unified_json = []
         if not game_details["dlcs"]:
             return []
+
+        ### FIXME: URL doesn't work with more than 50 product id's
         all_products_url = game_details["dlcs"]["expanded_all_products_url"]
-        return self.make_api_request(all_products_url)
+
+        # If ids greater than 50 split request
+        if game_details["dlcs"]["products"].size() > 50:
```
Comment (ItsAllAboutTheCode) [inline lutris/services/gog.py]: A constant should be used to better document the number of product IDs instead of a plain number such as 50.
i.e 
```python
MAX_PRODUCT_IDS_PER_REQUEST = 50
```

File: lutris/services/gog.py [hunk 2]
```diff
@@ -282,10 +282,51 @@ def get_products_page(self, page: int = 1, search: str | None = None) -> Any:
...
+            # Get url prefix and suffix, excluding product ids list
+            all_products_url_prefix = all_products_url[: (all_products_url.find("ids=") + 4)]
+            all_products_url_suffix = all_products_url[(all_products_url.find("\u0026")): -1]
```
Comment (ItsAllAboutTheCode) [inline lutris/services/gog.py]: The check for an ampersand here hurts readability  by other developers by searching for hex 0x26. Instead this should just check for the literal `&` character.

File: lutris/services/gog.py [hunk 3]
```diff
@@ -282,10 +282,51 @@ def get_products_page(self, page: int = 1, search: str | None = None) -> Any:
     def get_game_dlcs(self, product_id: str) -> list[dict]:
         """Return the list of DLC products for a game"""
         game_details = self.get_game_details(product_id)
+        unified_json = []
         if not game_details["dlcs"]:
             return []
+
+        ### FIXME: URL doesn't work with more than 50 product id's
```
Comment (ItsAllAboutTheCode) [inline lutris/services/gog.py]: Instead of having a FIXME here, you can reference the GOG api documentation that mentions that the product IDs have a max count of 50 from the following URL: https://gogapidocs.readthedocs.io/en/latest/galaxy.html#get--products

File: lutris/services/gog.py [hunk 4]
```diff
@@ -282,10 +282,51 @@ def get_products_page(self, page: int = 1, search: str | None = None) -> Any:
...
         all_products_url = game_details["dlcs"]["expanded_all_products_url"]
-        return self.make_api_request(all_products_url)
+
+        # If ids greater than 50 split request
+        if game_details["dlcs"]["products"].size() > 50:
+            # Get url prefix and suffix, excluding product ids list
+            all_products_url_prefix = all_products_url[: (all_products_url.find("ids=") + 4)]
```
Comment (ItsAllAboutTheCode) [inline lutris/services/gog.py]: If the product url doesn't have a "ids=", this results in a expression of `all_products_url[:3]`, which would remove the first 3 characters in the URL.
This should perform an if check first to determine if the `ids` option is part of the URL.
I also recommended storing the "ids=" in a constant and then performing taking the length of that constant when getting the size of the match instead of using `+ 4`. (Minor protection if the "ids=" code has to be updated in the future to check a different option).
```suggestion
            ID_PREFIX = "ids="
            if id_index := all_products_url.find(ID_PREFIX):
                all_products_url_prefix = all_products_url[: id_index + len(ID_PREFIX))]
```

File: lutris/services/gog.py (hunk 1 above)
Comment (ItsAllAboutTheCode) [inline lutris/services/gog.py]: For hardening  purposes, a check to make sure that the products key actually exist under the dlc can be performed to prevent exceptions in unusual circumstances.
i.e
```python
dlc_products = game_details["dlcs"].get("products")
if dlc_products and dlc_products.size() > 50:
```

File: lutris/services/gog.py [hunk 5]
```diff
@@ -282,10 +282,51 @@ def get_products_page(self, page: int = 1, search: str | None = None) -> Any:
...
+                try:
+                    start_index = 50 * i
+                    # Check if start index is out of bounds
+                    if start_index >= len(product_ids):
+                        break
+                    # Check if end index is out of bounds
+                    if start_index + 49 >= len(product_ids):
+                        end_index = -1
+                    else:
+                        end_index = start_index + 49
```
Comment (ItsAllAboutTheCode) [inline lutris/services/gog.py]: You can simplify the logic a bit by using a step size of `MAX_PRODUCT_IDS_PER_REQUEST` and then using the python `min` function to clamp the end_index to be no larger than the length of the product IDs array.
FYI: the `//` operator can be used to perform integer division, so the conversion to `int` isn't needed
```suggestion
            for start_index in range(0, len(product_ids), MAX_PRODUCT_IDS_PER_REQUEST):
                try:
                    end_index = max(start_index + MAX_PRODUCT_IDS_PER_REQUEST, len(product_ids))
```

File: tests/services/_test_gog.py [hunk 6]
```diff
@@ -0,0 +1,319 @@
...
+                        "total_size": 200,
+                        "files": [{"downlink": "https://example.invalid/dlc/soundtrack.zip"}],
+                    }
+                ]
+            },
+        }
+    ]
+
+    monkeypatch.setattr(GOGService, "get_game_details", lambda self, gogid: game)
+    monkeypatch.setattr(GOGService, "get_game_dlcs", lambda self, gogid: dlcs)
```
Comment (ItsAllAboutTheCode) [inline tests/services/_test_gog.py]: It would be good to have an edge case test for when the DLC product array is empty

**Comment by danieljohnson2** [comment]
> Either way, after you setup the development dependencies, you will need to [make format](https://github.com/lutris/lutris/blob/master/Makefile#L110) to fix the ruff formatting issues with your code.

Maybe not *need*- it's nice if you can format your code for us, but I'll merge it and format it myself if it comes to that. I think @ItsAllAboutTheCode's review comments are the main thing you need to look at.

File: lutris/services/gog.py [hunk 7]
```diff
@@ -84,6 +84,8 @@ class GOGService(OnlineService):
     medias = {"banner_small": GogSmallBanner, "banner": GogMediumBanner, "banner_large": GogLargeBanner}
     default_format = "banner"
 
+    MAX_PRODUCT_IDS_PER_REQUEST = 50
```
Comment (ItsAllAboutTheCode) [inline lutris/services/gog.py]: nitpick: This doesn't need to be a class member and can be put at module scope outside of any class

File: lutris/services/gog.py [hunk 8]
```diff
@@ -282,10 +284,50 @@ def get_products_page(self, page: int = 1, search: str | None = None) -> Any:
...
+        if game_details["dlcs"]["products"].size() > self.MAX_PRODUCT_IDS_PER_REQUEST:
+            # Get product ids only form json
+            product_ids = []
+            for product in game_details["dlcs"]["products"]:
+                product_ids.append(product["id"])
+            # Get number of requests by batches of 50
+            if len(product_ids) % self.MAX_PRODUCT_IDS_PER_REQUEST <= 0:
+                num_requests = len(product_ids) / self.MAX_PRODUCT_IDS_PER_REQUEST
+            else:
+                num_requests = (len(product_ids) / self.MAX_PRODUCT_IDS_PER_REQUEST) + 1
```
Comment (ItsAllAboutTheCode) [inline lutris/services/gog.py]: a modulo (%) can never return a negative remainder, so the <= check should be == 0.

However this could be simplified to using `math.ceil` to round to the nearest integer and avoid the need of the modulo



```suggestion
            math.ceil(len(product_ids) / self.MAX_PRODUCT_IDS_PER_REQUEST)```

File: lutris/services/gog.py [hunk 9]
```diff
@@ -282,10 +284,50 @@ def get_products_page(self, page: int = 1, search: str | None = None) -> Any:
...
             return []
-        all_products_url = game_details["dlcs"]["expanded_all_products_url"]
-        return self.make_api_request(all_products_url)
+        ### GOG Api limits requests to 50 comma-separated ids: https://gogapidocs.readthedocs.io/en/latest/galaxy.html#get--products
+        # If ids greater than 50 split request
+        if game_details["dlcs"]["products"].size() > self.MAX_PRODUCT_IDS_PER_REQUEST:
+            # Get product ids only form json
+            product_ids = []
+            for product in game_details["dlcs"]["products"]:
+                product_ids.append(product["id"])
```
Comment (ItsAllAboutTheCode) [inline lutris/services/gog.py]: This still needs to check that the `product` object actually has an "id" key to prevent an exception if it doesn't.
```suggestion
                product_id = product.get("id")
                if not product_id:
                    continue;
                product_ids.append(product_id)
```

File: tests/services/_test_gog.py [hunk 10]
```diff
@@ -0,0 +1,319 @@
...
+
+    spec = importlib.util.spec_from_file_location("lutris.services.gog", GOG_PATH)
+    module = importlib.util.module_from_spec(spec)
+    sys.modules["lutris.services.gog"] = module
+    assert spec.loader is not None
+    spec.loader.exec_module(module)
+    return module
+
+
+GOG = _install_stubs()
```
Comment (ItsAllAboutTheCode) [inline tests/services/_test_gog.py]: What is with all the manual module creation?
Why can't regular imports be used?

If it is a required a comment needs to be added to explain why.

This kind of manipulation of the sys.modules can break unit test depending on module import order as seen here https://github.com/lutris/lutris/commit/80e6c46ee8d0c59a41fa7acdd4d256f0db102dd6

File: lutris/services/gog.py (hunk 8 above)
Comment (ItsAllAboutTheCode) [reply lutris/services/gog.py]: Nevermind, this has been resolved to due changing the for loop iteration variable.

File: tests/services/_test_gog.py [hunk 11]
```diff
@@ -0,0 +1,319 @@
...
+                        "type": "mp3",
+                        "total_size": 200,
+                        "files": [{"downlink": "https://example.invalid/dlc/soundtrack.zip"}],
+                    }
+                ]
+            },
+        }
+    ]
+
+    monkeypatch.setattr(GOGService, "get_game_details", lambda self, gogid: game)
```
Comment (ItsAllAboutTheCode) [inline tests/services/_test_gog.py]: If you need to do any kind of patching you can use the `unittest.mock.Mock` class wraps functionality: https://docs.python.org/3/library/unittest.mock.html#unittest.mock.Mock

And then use the `with` context manager to patch a specific block of code

```python
from lutris.services.gog import GOGService
from unittest.mock import MagicMock, patch
# ...

mock_gog_service = MagicMock(wraps=GOGService)
mock_gog_service.get_game_details.side_effect = lambda self, gogid: game)
mock_gog_service.get_game_dlc.side_effect = lambda self, gogid: dlc)

# Patches the GOGService.get_game_details and GOGService.get_game_dlc to invoke mock methods.
# All other methods passthru to the normal GOGService impl
with patch("lutris.services.gog.GOGService", mock_gog_service) as _mock_gog_serv:
   service = GOGService()
   extras = service.get_extras("1344230395")
```

**Comment by KTGX710** [comment]
Thank you a ton @ItsAllAboutTheCode, really. All your guidance has been a huge help. I feel like this version is better than ever and I couldn't have done it without all the help. I'm still learning a lot, this is my first work contributing to an open-source platform, and I really appreciate you taking the time to walk me through all my errors.
I hope this version holds up to the teams' standards and that I can continue to work with everyone in the future building up Lutris.

P.S. Sorry about all the messy commits, I got a bit trigger happy with making sure I kept my repositories in sync.

**Comment by ItsAllAboutTheCode** [comment]
> Thank you a ton @ItsAllAboutTheCode, really. All your guidance has been a huge help. I feel like this version is better than ever and I couldn't have done it without all the help. I'm still learning a lot, this is my first work contributing to an open-source platform, and I really appreciate you taking the time to walk me through all my errors. I hope this version holds up to the teams' standards and that I can continue to work with everyone in the future building up Lutris.

No problem, if you need any guidance/help with making any of the changes or have any discussion points for the PR on just let me know.
I am just another developer trying to contribute to the project.

**Comment by KTGX710** [comment]
> > Thank you a ton @ItsAllAboutTheCode, really. All your guidance has been a huge help. I feel like this version is better than ever and I couldn't have done it without all the help. I'm still learning a lot, this is my first work contributing to an open-source platform, and I really appreciate you taking the time to walk me through all my errors. I hope this version holds up to the teams' standards and that I can continue to work with everyone in the future building up Lutris.
> 
> No problem, if you need any guidance/help with making any of the changes or have any discussion points for the PR on just let me know. I am just another developer trying to contribute to the project.

I'll definitely take all the guidance I can get. In my daytime job, I've learned that colaboration and getting others' perspective on problems is really important and will only help to build better solutions. So to say again, I really appreciate you taking the time to critique my work. I hope the current version implements everything well.

Being a Comp-Sci student only goes so far in learning to develop good code, so I wanted to start early in trying to build real solutions. While being a professional software developer may not be my end career goal, I would like to contribute to open-source platforms like this as much as my time allows in the future.

**Comment by danieljohnson2** [comment]
Well, the excitement seems to have died down, and we're heading for a release, so I think I better merge this now.
